src/java/org/apache/cassandra/cql3/statements/TuneUpReplicationFactor.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.cql3.statements;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import com.google.common.collect.ImmutableSet;
+import org.slf4j.Logger;
+
+import org.apache.cassandra.config.SchemaConstants;
+import org.apache.cassandra.locator.NetworkTopologyStrategy;
+import org.apache.cassandra.locator.SimpleStrategy;
+import org.apache.cassandra.schema.KeyspaceParams;
+import org.apache.cassandra.schema.ReplicationParams;
+import org.apache.cassandra.service.ClientWarn;
+
+
+public class TuneUpReplicationFactor
+{
+    private static final Set<String> SYSTEM_KEYSPACES = ImmutableSet.copyOf(Arrays.asList(
+                                                                            SchemaConstants.SYSTEM_KEYSPACE_NAME,
+                                                                            SchemaConstants.AUTH_KEYSPACE_NAME,
+                                                                            SchemaConstants.DISTRIBUTED_KEYSPACE_NAME,
+                                                                            SchemaConstants.TRACE_KEYSPACE_NAME,
+                                                                            SchemaConstants.SCHEMA_KEYSPACE_NAME)
+    );
+    public static final String REPLICATION_FACTOR = "replication_factor";
+
+    public final static TuneUpReplicationFactor instance = new TuneUpReplicationFactor();
+    private final AtomicBoolean uptuningEnabled = new AtomicBoolean(true);
+
+    public boolean isUptuningEnabled()
+    {
+        return uptuningEnabled.get();
+    }
+
+    public void setUptuningEnabled(boolean uptuningEnabled)
+    {
+        this.uptuningEnabled.set(uptuningEnabled);
+    }
+
+
+    public Map<String, String> apply(Logger logger, String keyspaceName, KeyspaceParams inputKeyspaceParams)
+    {
+        final Set<String> clientWarnings = new HashSet<>();
+        final Map<String, String> replicationParams = inputKeyspaceParams.replication.asMap(); // new hashmap just created for client use, ok to modify
```

Comment:
I could, but I don't see advantage in doing that (and IIRC, for Cassandra 4 the interfaces are a bit different so I would make the two implementations diverge). 

We cannot alter ReplicationParams instances directly; so I'd still need to use an intermediate data structure for manipulation, and, nts() constructor would require a bit of a cumbersome composition (since it needs an array). Here we bring a map in, we push a map out. Surely ReplicationParams->ReplicationParams would work out as well but.. no advantage, IMHO.

Reply:
(I thought I had written a response but I must have forgotten to click)

The real reason for this is that, this way, we don't have to handle all possible strategies; we just ignore what we don't know. Otherwise we'd need a code path just to rebuild a specific ReplicationParams instance (including local, and possibly other strategies if they ever add them, or if anybody implements a custom strategy). So, it seems the least-resistance path, and doesn't seem to have real problems.

Reply:
ah, the `nts(` function does look cumbersome. And seems to be used only in tests... Okay, let's keep the map manipulation. As to different code paths - but you could return the existing object unmodified for the strategies we don't know, or I'm missing something? Basically replace the "if map.get(CLASS)" with "if params.klass == ...".

Reply:
@dmitry-potepalov we could, but, what is the difference in terms of I/O? We don't change the params right now if we don't know the strategy. I'm not sure what worries you; try telling me what you deem to be the risky/error-prone part.

Reply:
@dmitry-potepalov ping (just in case you missed my question)